doc/src/build/programming-with-objects/ch5-dynamic-fields.md

```diff
@@ -0,0 +1,166 @@
+---
+title: Chapter 5 - Dynamic Fields
+---
+
+In previous chapters, we walked through various ways to use object fields to store primitive data and other objects (wrapping), but there are a few limitations to this approach:
+
+1. Object's have a finite set of fields keyed by identifiers that are fixed when its module is published.
+2. An object can become very large if it wraps several other objects. Larger objects can lead to higher gas fees in transactions. In addition, there is an upper bound on object size.
+3. As we will see in future chapters, there will be use cases where we need to store a collection of objects of heterogeneous types. Since the Move `vector` type must be templated on one single type `T`, it is not suitable for this.
+
+Fortunately, Sui provides *dynamic fields* with arbitrary names (not just identifiers), added and removed on-the-fly (not fixed at publish), which only affect gas when they are accessed, and can store heterogeneous values. This chapter introduces the libraries for interacting with this kind of field.
+
+### Fields vs Object Fields
+
+There are two flavors of dynamic field -- "fields" and "object fields" -- which differ based on how their values are stored:
+
+- **Object field** values *must* be objects themselves (have the `key` ability, and `id: UID` as the first field) and will still appear under their ID to external tools (explorers, wallets, etc) accessing storage even when stored in a dynamic object field.
+- **Field** values do not need to be objects. An object stored in a (non-object) field will not be accessible via its ID while it is stored in a dynamic field.
```

Comment:
nit, flip there order here to be consistent. I know it will require some rewording, but it is strange to go "Fields vs Object Fields" and then "fields and object fields", but then list "object field" first.

Reply:
Yeah, you're right... for whatever reason, it was easier for me to explain object fields first, but I seem to have gotten over that hang-up in the second draft!